control_toolbox/include/control_toolbox/pid.hpp

```diff
@@ -39,64 +39,151 @@
 #include <limits>
 #include <string>
 
+#include "fmt/format.h"
 #include "rclcpp/duration.hpp"
 #include "realtime_tools/realtime_buffer.hpp"
 
 namespace control_toolbox
 {
-class AntiwindupStrategy
+/**
+ * \brief Antiwindup strategy for PID controllers.
+ *
+ * This class defines various antiwindup strategies that can be used in PID controllers.
+ * It allows setting the type of antiwindup strategy and validates the parameters accordingly.
+ *
+ * \param i_max Upper integral clamp.
+ * \param i_min Lower integral clamp.
+ * \param u_max Upper output clamp.
+ * \param u_min Lower output clamp.
+ * \param tracking_time_constant Specifies the tracking time constant for the 'back_calculation' strategy. If set
+ *    to 0.0 when this strategy is selected, a recommended default value will be applied.
+ * \param legacy_antiwindup Anti-windup functionality. When set to true, limits
+      the integral error to prevent windup; otherwise, constrains the
+      integral contribution to the control output. i_max and
+      i_min are applied in both scenarios.
+  * \param error_deadband Error deadband is used to avoid integration when the error is within a small range.
+  * \param type Specifies the antiwindup strategy type. Valid values are:
+ *   - `NONE`: No antiwindup strategy applied.
+ *   - `LEGACY`: Legacy antiwindup strategy, which limits the integral term to prevent windup.
+ *   - `BACK_CALCULATION`: Back calculation antiwindup strategy, which uses a tracking time constant.
+ *   - `CONDITIONAL_INTEGRATION`: Conditional integration antiwindup strategy, which integrates only when certain conditions are met.
+ */
+struct AntiWindupStrategy
 {
 public:
   enum Value : int8_t
   {
-    NONE = 0,
+    UNDEFINED = -1,
+    NONE,
+    LEGACY,
     BACK_CALCULATION,
     CONDITIONAL_INTEGRATION
   };
 
-  constexpr AntiwindupStrategy() : value_(NONE) {}
-  constexpr AntiwindupStrategy(Value v) : value_(v) {}  // NOLINT(runtime/explicit)
+  AntiWindupStrategy()
+  : type(UNDEFINED),
+    i_min(std::numeric_limits<double>::quiet_NaN()),
+    i_max(std::numeric_limits<double>::quiet_NaN()),
+    legacy_antiwindup(false),
+    tracking_time_constant(0.0),
+    error_deadband(std::numeric_limits<double>::epsilon())
+  {
+  }
 
-  explicit AntiwindupStrategy(const std::string & s)
+  void set_type(const std::string & s)
   {
     if (s == "back_calculation")
     {
-      value_ = BACK_CALCULATION;
+      type = BACK_CALCULATION;
     }
     else if (s == "conditional_integration")
     {
-      value_ = CONDITIONAL_INTEGRATION;
+      type = CONDITIONAL_INTEGRATION;
+    }
+    else if (s == "legacy")
+    {
+      type = LEGACY;
```

Comment:
should we add a deprecation notice here?

Reply:
Done